### ISSUE TITLE
Notes from Meeting 2026-01-21

### ISSUE THREAD

**Issue by willGraham01** [UCL/causalprog#122]
We should spend the remaining time focusing on a couple of families of graphs, that have the following layout:

```bash
    U_X ->  U_Y
     |       |
Z -> X  ->   Y

     U --->---
     |       |
     |       |
Z -> X  ->   Y
```

The graph is also parametrised by some set of parameters $\theta$.
The $U, U_X, U_Y$ represent the "hidden" effects that we can't observe.
$Z$, $X$, and $Y$ represent RVs that we can observe.
The $\theta$ parameters may determine (parts of the) forms of the RVs in the model too.

> The graphs are something we can already set up.
> The distinction is that the graphs above don't draw the model parameters as nodes too, and link them to the RVs that use them, whereas we require our graphs to do this.
> That is, suppose that there is a parameter $\mu$ that defines the structure of the RV $X$.
> `causalprog` requires a node to be placed that represents $\mu$, and then have this node linked to $X$'s node.
> By contrast, the $\mu$ node and corresponding link to $X$ is absent from the graphs drawn above.

The graph itself is our model $M = M_\theta$.

> Again, `causalprog.graph.Graph` has a `.build_model` method for this reason.
> This method returns a $\theta$-parametrised function, which when evaluated returns a Python object can then draw samples from the corresponding $M_{\theta}$.
> That is, `Graph.build_model()(theta)` "equals" $M_{\theta}$.

The model can spit out predictions for certain quantities (for example, $ \mathbb{E}_{M_\theta} [Y | x_i, z_i ] $ ).
Such predictions are in actual fact deterministic functions of $\theta$ (that is there exists some function $g_{\theta}$ such that $ g_{\theta}(x_i, z_i) = \mathbb{E}_{M_\theta} [Y | x_i, z_i ] $ ).
One can also have a naive or black-box model $L$ predict the same model quantities (such predictions would be written as, e.g., $ \mathbb{E}_{L} [Y | x_i, z_i ] $ ).

> This is where external `pytorch` stuff enters into the picture.
> One could feed some data into a `torch` model to train it, and then have the resulting trained model $L$ make predictions given the $x_i$ and $z_i$.

> This is also my first point of confusion for `causalprog` implications; is $L$ actually $L_{\theta}$?
> Namely are the $\phi_i^L$ really just supplied as static vectors; _or_ is some black-box model $L_{\theta}$ provided which, given a particular value for $\theta$, makes predictions for $\phi_i^L$ (in which case, $L$ needs to be suitably nice programmatically to allow for differentiation etc)?
> I'm increasingly thinking that it's the latter (not static), since otherwise there aren't really any concerns for how we build the `pytorch` models - they are effectively constants by the time they hit the causal problem and consequentially `causalprog`?
> It would also give more context to the talk of normalising flows (which are known to have all these nice properties we need to do numerical solving on them).

A causal problem then defines a causal estimand $\sigma = \sigma_{\theta}$, which is some quantity for which we want to estimate upper and lower bounds for, as predicted by the model $M_{\theta}$.
We also define a number of constraint functions $\phi_i^{\theta}$ (indexed by $i$), which again are quantities from the model $M_{\theta}$, but which we are also able to estimate using the black-box model $L$.
The predictions that $L$ makes for the values of $\phi_i^{\theta}$ we denote by $\phi_i^L$.
The causal problem to solve is then;

$$ \mathrm{min/max}_{\theta} \sigma_{\theta}, \quad \text{subject to } \vert\vert \phi_i^{\theta} - \phi_i^L \vert\vert^2 \leq \epsilon $$

for some given tolerance $\epsilon$ and suitable (vector) norm $\vert\vert\cdot\vert\vert$.
The constraints may also be represented as some vector $\Phi^{\theta}$ too, in which case a vector norm would be used.
The $\phi_i^L$ are fixed (non-$\theta$-dependant) values that are computed a-priori by some suitable external model, so are effectively constants.
At this point, Ricardo said that most of these problems would be solved via Augmented Lagrangian methods.

> `causalprog.causal_problem.CausalProblem` has a `.lagrangian` method for this purpose too.

Regardless, $\sigma_{\theta}$ is always required to be estimated using $M_{\theta}$.
Ricardo mentioned that one of the most common forms for $\sigma$ is

$$ \sigma_{\theta} = \mathbb{E}_{M_{\theta}}[Y \vert \mathrm{do}(X = x)], $$

which is what we'll focus on for the rest of the package's development.
Evaluation of this quantity will typically require some Monte-Carlo integral evaluation, drawing from appropriate distributions for the sampling, so will require us to be clever about how we go about constructing and evaluating it (so the solver doesn't slow down too much)!

The $\phi_i^{\theta}$ are also model quantities, that need to be evaluated via $M_{\theta}$.
Going with the example above and from Ricardo's notes, these might be of the form

$$ \phi_i^{\theta} = \mathbb{E}_{M_{\theta}}[Y \vert x_i, z_i]. $$

This means that the $\phi_i^L$ would be

$$ \phi_i^L = \mathbb{E}_{L}[Y \vert x_i, z_i], $$

using the black-box model $L$.

## What do we already have?

- We can build the graphs.
  - Our nodes currently only support scalar-valued RVs, but in general the nodes would need to be vectors.
- The `Constraint` class effectively already represents the constraints, but it needs to be given:
  - The pre-computed $\phi_i^L$ values (as the `data` argument to the constructor).
    Note that if the $L = L_{\theta}$ hunch is correct, we'll need to slightly adjust this class to accept this argument, but it should just be a like-for-like substitution.
  - $\epsilon$ (as the `tolerance` argument to the constructor).
  - Functions $f_i$ for evaluating $\phi_i^{\theta}$ given _sample values from the model_ $M_{\theta}$ (as the `model_quantity` argument).
- The `CausalEstimand` class represents $\sigma_{\theta}$, but needs to be given:
  - A function $g$ for evaluating $\sigma_{\theta}$ given _sample values from the model_ $M_{\theta}$.
- The `CausalProblem` class then provides utility methods for assembly of the Lagrangian for either the maximisation or minimisation problem that needs to be solved.
  It can also handle applications of the "do" operations and similar via effect handlers.

As such, we actually have all the end-to-end components we need provided that the user can define:

- An efficient and clever way for a user to specify and/or define the $f_i$ and $g$ functions.
  These will typically be Monte-Carlo integrals that need to be evaluated, which is why it is important they assume samples are available from $M_{\theta}$.
  It also means that each evaluation of these functions will require re-generating appropriate samples given $\theta$ from the graph model $M_{\theta}$.
- Their own model that provides the estimations for $\phi_i^L$.

The above, plus making the package more friendly for working with the specific family of problems that Ricardo was describing (above), is what we'll be focusing on going forward.

## Example concerning the "do" operator

Using the $Z, X, U_X, U_Y, Y$ graph structure above, let's assume further that $U_x, U_y \sim \mathcal{N}(0, \nu^2)$.
If $X = aZ^2 + bZ + U_x$, and $Y = cX^2 + dX + U_Y$, then we expect that

$$ \sigma_{\theta} := \mathbb{E}_{M_{\theta}}[Y \vert \mathrm{do}(X = x)] = cx^2 + dx, $$

however this is distinct from

$$ \phi_i^{\theta} := \mathbb{E}_{M_{\theta}}[Y \vert x_i, z_i] = cx_i^2 + dx_i + \mathbb{E}_{M_{\theta}}[U_Y|x_i, z_i]. $$

If we were to translate this into `causalprog` terms above, we'd need to provide

- $g_{\theta}$; when provided samples from $M_{\theta}$ subject to the effect handler " $\mathrm{do}(X = x)$ ", returns the value of $ \mathbb{E}_{M_{\theta}}[Y \vert \mathrm{do}(X = x)] $.
- $f_i$; when provided samples from $M_{\theta}$ subject to the effect handler "given $X=x_i$ and $Z=z_i$", returns the value of $\mathbb{E}_{M_{\theta}}[Y \vert x_i, z_i]$.
- Some $\phi_i^L$ predictions from an outside, black-box model.
  (Or, in the event that $L = L_{\theta}$, a function that given $\theta$ returns $\phi_i^L$).

**Comment by mscroggs** [comment]
Closing this as I think any information in here that is still relevant is included in docs following #172
